### PR TITLE
[SPARK-13253] [SQL]avoid checking nullability for complex data type for typeSuffix code path

### PR DESCRIPTION
**Pull request by kevinyu98**
Hello: 

When we call the typeSuffix method, it will call the dataType and get the LongType for the suffix("L"). But in the complex data type(like CreatArray) cases, the dataType will also evaluate the children's nullability, which is not necessary for the typeSuffix. 

For the proposed fix, I will create a prettyDataType, in parallel to the dataType in the expression. At the base, it defaults to the dataType. It is only used by typeSuffix in the NameExpression for now.

So the main changes is at the typeSuffix in the NameExpression, and the complexTypeCreator. The rest of files just override the prettyDataType from the abstract class Expression when their "dataType" method relies on other expression's "dataType"

Then for those complex types, the "prettyDataType" does not try to evaluate the "nullable" but just pass a default that will not be used at all by the caller, "typeSuffix".

### CODE REVIEW

File: python/pyspark/sql/functions.py [hunk 1]
```diff
@@ -250,17 +250,46 @@ def corr(col1, col2):
...
+    >>> b = [2 * x for x in range(20)]
+    >>> df = sqlContext.createDataFrame(zip(a, b), ["a", "b"])
+    >>> df.agg(corr("a", "b").alias('c')).collect()
+    [Row(c=1.0)]
     """
     sc = SparkContext._active_spark_context
     return Column(sc._jvm.functions.corr(_to_java_column(col1), _to_java_column(col2)))
 
 
+@since(2.0)
```
Comment (frreiss) [inline python/pyspark/sql/functions.py]: These changes should probably go into a different PR.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/namedExpressions.scala [hunk 2]
```diff
@@ -87,7 +87,7 @@ trait NamedExpression extends Expression {
 
   protected def typeSuffix =
     if (resolved) {
-      dataType match {
+      prettyDataType match {
```
Comment (frreiss) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/namedExpressions.scala]: This line appears to be the only code that consumes the new `prettyDataType` attribute

**Comment by frreiss** [comment]
This change adds a new attribute `prettyDataType` to about two dozen different classes. The new attribute is populated with code that is mostly copied and pasted from the code to set the existing `dataType` attribute. The copied code has been modified to set all "nullable" attributes in nested types to TRUE. As far as I can see, the `prettyDataType` attribute is only used in one place.
It would be cleaner to compute `prettyDataType` in a single function by walking a visitor over the tree stored in the `dataType` field.

**Comment by yzhou2001** [comment]
I'm not sure whether the visitor pattern is a good choice here. First introduction of a visitor just for this fix would see no less code changes; Second Scala pattern matching capabilities notwithstanding, it also boils down to the model hierarchy stability vs. capability stability. Pattern matching could have been utilized here if not for concerns on maintaining the original exception throwing from the "dataType" method.

**Comment by kevinyu98** [comment]
@marmbrus : Hello Michael, can you help review this PR? Let me know anything I need to change. Thank very much ! Kevin

**Comment by kevinyu98** [comment]
@marmbrus : sorry, just saw Wenchen Fan's update from jira, the problem has been solved. I will close this PR.
